// sign-pdf-lib/dist/src/signer/pdf-document-digital-signer.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const pdf_document_digital_signer_1 = require("./pdf-document-digital-signer");
const errors_1 = require("../errors");
const _helpers_1 = require("../../test/_helpers");
const assets_pdf_document_digital_signer_1 = require("../../test/_run-assets/signer/assets-pdf-document-digital-signer");
const chai_1 = require("chai");
const pdf_lib_1 = require("pdf-lib");
it('_generate', function () {
    return __awaiter(this, void 0, void 0, function* () {
        const pdf = yield (0, _helpers_1.generatePdfAsync)({ pageCount: 2 });
        yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.pdf, pdf);
        const placeholderPdf = yield (0, _helpers_1.generatePlaceholderPdfAsync)(pdf);
        yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.realPlaceholderPdf, placeholderPdf);
        const fieldPdf = yield (0, _helpers_1.generateFieldPdfAsync)(pdf);
        yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.realFieldPdf, fieldPdf);
    });
});
describe('PdfDocumentDigitalSigner', function () {
    let pdfDocSigner;
    let addFieldParameters;
    let addVisualParameters;
    let addPlaceholderParameters;
    let updateParameters;
    beforeEach(function () {
        return __awaiter(this, void 0, void 0, function* () {
            addFieldParameters = {
                name: 'Signature',
                pageIndex: 0,
                rectangle: {
                    left: 50,
                    top: 100,
                    right: 50 + 214,
                    bottom: 100 + 70
                },
                visualRef: pdf_lib_1.PDFRef.of(170),
                placeholderRef: pdf_lib_1.PDFRef.of(171),
                embedFont: true
            };
            addVisualParameters = {
                background: assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.visual,
                texts: [
                    {
                        lines: [
                            'JOHN',
                            'DOE'
                        ]
                    }, {
                        lines: [
                            'Digitally signed by',
                            'JOHN DOE',
                            'Date: 2023.11.03',
                            '20:28:46 +02\'00\''
                        ]
                    }
                ]
            };
            addPlaceholderParameters = {
                name: 'Test Signer',
                location: 'Timisoara',
                reason: 'Signing',
                date: new Date(2023, 1, 20, 18, 47, 35),
                contactInfo: 'signer@semnezonline.ro',
                signaturePlaceholder: 'A'.repeat(4096),
                rangePlaceHolder: 9999
            };
            updateParameters = {
                placeholderRef: pdf_lib_1.PDFRef.of(180),
                visualRef: pdf_lib_1.PDFRef.of(181),
                embedFont: true
            };
            pdfDocSigner = yield pdf_document_digital_signer_1.PdfDocumentDigitalSigner.fromPdfAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.pdf);
        });
    });
    describe('addSignatureField', function () {
        it('adds signature field', function () {
            return __awaiter(this, void 0, void 0, function* () {
                pdfDocSigner.addSignatureField(addFieldParameters);
                const fieldPdf = yield pdfDocSigner.saveAsync();
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.fieldPdf, fieldPdf);
                (0, chai_1.expect)(fieldPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.fieldPdf);
            });
        });
        it('adds signature field (no name)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete addFieldParameters.name;
                pdfDocSigner.addSignatureField(addFieldParameters);
                const noNameFieldPdf = yield pdfDocSigner.saveAsync();
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noNameFieldPdf, noNameFieldPdf);
                (0, chai_1.expect)(noNameFieldPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noNameFieldPdf);
            });
        });
        it('adds signature field (no rectangle)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete addFieldParameters.rectangle;
                pdfDocSigner.addSignatureField(addFieldParameters);
                const noRectangleFieldPdf = yield pdfDocSigner.saveAsync();
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noRectangleFieldPdf, noRectangleFieldPdf);
                (0, chai_1.expect)(noRectangleFieldPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noRectangleFieldPdf);
            });
        });
        it('adds signature field (no visual ref)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete addFieldParameters.visualRef;
                pdfDocSigner.addSignatureField(addFieldParameters);
                const noVisualRefFieldPdf = yield pdfDocSigner.saveAsync();
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noVisualRefFieldPdf, noVisualRefFieldPdf);
                (0, chai_1.expect)(noVisualRefFieldPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noVisualRefFieldPdf);
            });
        });
        it('adds signature field (no placeholder ref)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete addFieldParameters.placeholderRef;
                pdfDocSigner.addSignatureField(addFieldParameters);
                const noPlaceholderRefFieldPdf = yield pdfDocSigner.saveAsync();
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noPlaceholderRefFieldPdf, noPlaceholderRefFieldPdf);
                (0, chai_1.expect)(noPlaceholderRefFieldPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noPlaceholderRefFieldPdf);
            });
        });
        it('adds signature field (no font)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                addFieldParameters.embedFont = false;
                pdfDocSigner.addSignatureField(addFieldParameters);
                const noFontFieldPdf = yield pdfDocSigner.saveAsync();
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noFontFieldPdf, noFontFieldPdf);
                (0, chai_1.expect)(noFontFieldPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noFontFieldPdf);
            });
        });
        it('adds signature field (no optionals)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete addFieldParameters.name;
                delete addFieldParameters.rectangle;
                delete addFieldParameters.visualRef;
                delete addFieldParameters.placeholderRef;
                pdfDocSigner.addSignatureField(addFieldParameters);
                const noOptionalsFieldPdf = yield pdfDocSigner.saveAsync();
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noOptionalsFieldPdf, noOptionalsFieldPdf);
                (0, chai_1.expect)(noOptionalsFieldPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noOptionalsFieldPdf);
            });
        });
        it('adds signature field (page two)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                addFieldParameters.pageIndex = 1;
                pdfDocSigner.addSignatureField(addFieldParameters);
                const pageTwoFieldPdf = yield pdfDocSigner.saveAsync();
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.pageTwoFieldPdf, pageTwoFieldPdf);
                (0, chai_1.expect)(pageTwoFieldPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.pageTwoFieldPdf);
            });
        });
    });
    describe('addVisualAsync', function () {
        it('adds visual', function () {
            return __awaiter(this, void 0, void 0, function* () {
                const visualRef = yield pdfDocSigner.addVisualAsync(addVisualParameters);
                const visualPdf = yield pdfDocSigner.saveAsync();
                (0, chai_1.expect)(visualRef).to.not.be.undefined;
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.visualPdf, visualPdf);
                (0, chai_1.expect)(visualPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.visualPdf);
            });
        });
        it('adds visual (no background)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete addVisualParameters.background;
                const visualRef = yield pdfDocSigner.addVisualAsync(addVisualParameters);
                const noBackgroundVisualPdf = yield pdfDocSigner.saveAsync();
                (0, chai_1.expect)(visualRef).to.not.be.undefined;
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noBackgroundVisualPdf, noBackgroundVisualPdf);
                (0, chai_1.expect)(noBackgroundVisualPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noBackgroundVisualPdf);
            });
        });
        it('adds visual (no texts)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete addVisualParameters.texts;
                const visualRef = yield pdfDocSigner.addVisualAsync(addVisualParameters);
                const noTextsVisualPdf = yield pdfDocSigner.saveAsync();
                (0, chai_1.expect)(visualRef).to.not.be.undefined;
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noTextsVisualPdf, noTextsVisualPdf);
                (0, chai_1.expect)(noTextsVisualPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noTextsVisualPdf);
            });
        });
        it('adds visual (no optionals)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete addVisualParameters.background;
                delete addVisualParameters.texts;
                const visualRef = yield pdfDocSigner.addVisualAsync(addVisualParameters);
                const noOptionalsVisualPdf = yield pdfDocSigner.saveAsync();
                (0, chai_1.expect)(visualRef).to.be.undefined;
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noOptionalsVisualPdf, noOptionalsVisualPdf);
                (0, chai_1.expect)(noOptionalsVisualPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noOptionalsVisualPdf);
            });
        });
    });
    describe('addEmptyVisual', function () {
        it('adds empty visual', function () {
            return __awaiter(this, void 0, void 0, function* () {
                pdfDocSigner.addEmptyVisual();
                const emptyVisualPdf = yield pdfDocSigner.saveAsync();
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.emptyVisualPdf, emptyVisualPdf);
                (0, chai_1.expect)(emptyVisualPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.emptyVisualPdf);
            });
        });
    });
    describe('addSignaturePlaceholder', function () {
        it('adds placeholder', function () {
            return __awaiter(this, void 0, void 0, function* () {
                const placeholderRef = yield pdfDocSigner.addSignaturePlaceholder(addPlaceholderParameters);
                const placeholderPdf = yield pdfDocSigner.saveAsync();
                (0, chai_1.expect)(placeholderRef).to.not.be.undefined;
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.placeholderPdf, placeholderPdf);
                (0, chai_1.expect)(placeholderPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.placeholderPdf);
            });
        });
        it('adds placeholder (no name)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete addPlaceholderParameters.name;
                const placeholderRef = yield pdfDocSigner.addSignaturePlaceholder(addPlaceholderParameters);
                const noNamePlaceholderPdf = yield pdfDocSigner.saveAsync();
                (0, chai_1.expect)(placeholderRef).to.not.be.undefined;
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noNamePlaceholderPdf, noNamePlaceholderPdf);
                (0, chai_1.expect)(noNamePlaceholderPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noNamePlaceholderPdf);
            });
        });
        it('adds placeholder (no reason)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete addPlaceholderParameters.reason;
                const placeholderRef = yield pdfDocSigner.addSignaturePlaceholder(addPlaceholderParameters);
                const noReasonPlaceholderPdf = yield pdfDocSigner.saveAsync();
                (0, chai_1.expect)(placeholderRef).to.not.be.undefined;
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noReasonPlaceholderPdf, noReasonPlaceholderPdf);
                (0, chai_1.expect)(noReasonPlaceholderPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noReasonPlaceholderPdf);
            });
        });
        it('adds placeholder (no location)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete addPlaceholderParameters.location;
                const placeholderRef = yield pdfDocSigner.addSignaturePlaceholder(addPlaceholderParameters);
                const noLocationPlaceholderPdf = yield pdfDocSigner.saveAsync();
                (0, chai_1.expect)(placeholderRef).to.not.be.undefined;
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noLocationPlaceholderPdf, noLocationPlaceholderPdf);
                (0, chai_1.expect)(noLocationPlaceholderPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noLocationPlaceholderPdf);
            });
        });
        it('adds placeholder (no contact info)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete addPlaceholderParameters.contactInfo;
                const placeholderRef = yield pdfDocSigner.addSignaturePlaceholder(addPlaceholderParameters);
                const noContactInfoPlaceholderPdf = yield pdfDocSigner.saveAsync();
                (0, chai_1.expect)(placeholderRef).to.not.be.undefined;
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noContactInfoPlaceholderPdf, noContactInfoPlaceholderPdf);
                (0, chai_1.expect)(noContactInfoPlaceholderPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noContactInfoPlaceholderPdf);
            });
        });
        it('adds placeholder (no date)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete addPlaceholderParameters.date;
                const placeholderRef = yield pdfDocSigner.addSignaturePlaceholder(addPlaceholderParameters);
                const noDatePlaceholderPdf = yield pdfDocSigner.saveAsync();
                (0, chai_1.expect)(placeholderRef).to.not.be.undefined;
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noDatePlaceholderPdf, noDatePlaceholderPdf);
                (0, chai_1.expect)(noDatePlaceholderPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noDatePlaceholderPdf);
            });
        });
        it('adds placeholder (no optionals)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete addPlaceholderParameters.name;
                delete addPlaceholderParameters.reason;
                delete addPlaceholderParameters.location;
                delete addPlaceholderParameters.contactInfo;
                delete addPlaceholderParameters.date;
                const placeholderRef = yield pdfDocSigner.addSignaturePlaceholder(addPlaceholderParameters);
                const noOptionalsPlaceholderPdf = yield pdfDocSigner.saveAsync();
                (0, chai_1.expect)(placeholderRef).to.not.be.undefined;
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noOptionalsPlaceholderPdf, noOptionalsPlaceholderPdf);
                (0, chai_1.expect)(noOptionalsPlaceholderPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noOptionalsPlaceholderPdf);
            });
        });
        it('adds placeholder for different settings', function () {
            return __awaiter(this, void 0, void 0, function* () {
                addPlaceholderParameters.signaturePlaceholder = 'A'.repeat(5000),
                    addPlaceholderParameters.rangePlaceHolder = 9999;
                const placeholderRef = yield pdfDocSigner.addSignaturePlaceholder(addPlaceholderParameters);
                const differentSettingsPlaceholderPdf = yield pdfDocSigner.saveAsync();
                (0, chai_1.expect)(placeholderRef).to.not.be.undefined;
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.differentSettingsPlaceholderPdf, differentSettingsPlaceholderPdf);
                (0, chai_1.expect)(differentSettingsPlaceholderPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.differentSettingsPlaceholderPdf);
                (0, chai_1.expect)(differentSettingsPlaceholderPdf).to.not.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.placeholderPdf);
            });
        });
    });
    describe('updateSignature', function () {
        beforeEach(function () {
            return __awaiter(this, void 0, void 0, function* () {
                pdfDocSigner = yield pdf_document_digital_signer_1.PdfDocumentDigitalSigner.fromPdfAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.realFieldPdf);
            });
        });
        it('updates signature', function () {
            return __awaiter(this, void 0, void 0, function* () {
                pdfDocSigner.updateSignature('Signature', updateParameters);
                const updatePdf = yield pdfDocSigner.saveAsync();
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.updatePdf, updatePdf);
                (0, chai_1.expect)(updatePdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.updatePdf);
            });
        });
        it('updates signature (no visual ref)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete updateParameters.visualRef;
                pdfDocSigner.updateSignature('Signature', updateParameters);
                const noVisualRefUpdatePdf = yield pdfDocSigner.saveAsync();
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noVisualRefUpdatePdf, noVisualRefUpdatePdf);
                (0, chai_1.expect)(noVisualRefUpdatePdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noVisualRefUpdatePdf);
            });
        });
        it('updates signature (no font)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                updateParameters.embedFont = false;
                pdfDocSigner.updateSignature('Signature', updateParameters);
                const noFontUpdatePdf = yield pdfDocSigner.saveAsync();
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noFontUpdatePdf, noFontUpdatePdf);
                (0, chai_1.expect)(noFontUpdatePdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noFontUpdatePdf);
            });
        });
        it('updates signature (no optionals)', function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete updateParameters.visualRef;
                pdfDocSigner.updateSignature('Signature', updateParameters);
                const noOptionalsUpdatedPdf = yield pdfDocSigner.saveAsync();
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.noOptionalsUpdatedPdf, noOptionalsUpdatedPdf);
                (0, chai_1.expect)(noOptionalsUpdatedPdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.noOptionalsUpdatedPdf);
            });
        });
        it('throws for already signed pdf', function () {
            return __awaiter(this, void 0, void 0, function* () {
                pdfDocSigner = yield pdf_document_digital_signer_1.PdfDocumentDigitalSigner.fromPdfAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.updatePdf);
                (0, chai_1.expect)(() => pdfDocSigner.updateSignature('Signature', updateParameters)).to.throw(errors_1.AlreadySignedError);
            });
        });
    });
    describe('saveAsync', function () {
        it('saves pdf', function () {
            return __awaiter(this, void 0, void 0, function* () {
                const savePdf = yield pdfDocSigner.saveAsync();
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.savePdf, savePdf);
                (0, chai_1.expect)(savePdf).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.savePdf);
            });
        });
    });
    describe('getPlaceholderRanges', function () {
        it('returns placeholder ranges', function () {
            return __awaiter(this, void 0, void 0, function* () {
                pdfDocSigner = yield pdf_document_digital_signer_1.PdfDocumentDigitalSigner.fromPdfAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.realPlaceholderPdf);
                const ranges = yield pdfDocSigner.getPlaceholderRanges();
                yield _helpers_1.generateAsset.generateJsonAsync(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.paths.ranges, ranges);
                (0, chai_1.expect)(ranges).to.be.deep.equal(assets_pdf_document_digital_signer_1.pdfDocumentDigitalSignerAssets.ranges);
            });
        });
    });
});
