// sign-pdf-lib/dist/src/signer/pdf-visual-signer.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const pdf_visual_signer_1 = require("./pdf-visual-signer");
const assets_pdf_visual_signer_pdf_1 = require("../../test/_run-assets/signer/assets-pdf-visual-signer-pdf");
const _helpers_1 = require("../../test/_helpers");
const chai_1 = require("chai");
const chai_2 = require("chai");
const chaiAsPromised = require("chai-as-promised");
(0, chai_1.use)(chaiAsPromised);
it("_generate", function () {
    return __awaiter(this, void 0, void 0, function* () {
        const pdf = yield (0, _helpers_1.generatePdf17Async)({ pageCount: 2 });
        yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.paths.pdf, pdf);
        const reverseYPdf = yield (0, _helpers_1.generatePdf13Async)({ pageCount: 2 });
        yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.paths.reverseYPdf, reverseYPdf);
    });
});
describe("PdfVisualSigner", function () {
    let pdfSigner;
    let visualInfo;
    beforeEach(function () {
        visualInfo = {
            pageNumber: 1,
            rectangle: { left: 50, top: 100, right: 50 + 214, bottom: 100 + 70 },
            background: assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.signatureImage,
            texts: [
                {
                    lines: ["JOHN", "DOE"],
                },
                {
                    lines: [
                        "Digitally signed by",
                        "JOHN DOE",
                        "Date: 2023.11.03",
                        "20:28:46 +02'00'",
                    ],
                },
            ],
            // @ts-ignore
            backgroundName: "background1",
        };
        pdfSigner = new pdf_visual_signer_1.PdfVisualSigner();
    });
    describe("signAsync", function () {
        it("signs document", function () {
            return __awaiter(this, void 0, void 0, function* () {
                const visualSignedPdf = yield pdfSigner.signAsync(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.pdf, visualInfo);
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.paths.visualSignedPdf, visualSignedPdf);
                (0, chai_2.expect)(visualSignedPdf).to.be.deep.equal(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.visualSignedPdf);
            });
        });
        it("signs document (reverseY)", function () {
            return __awaiter(this, void 0, void 0, function* () {
                visualInfo.reverseY = true;
                const visualSignedReverseYPdf = yield pdfSigner.signAsync(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.reverseYPdf, visualInfo);
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.paths.visualSignedReverseYPdf, visualSignedReverseYPdf);
                (0, chai_2.expect)(visualSignedReverseYPdf).to.be.deep.equal(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.visualSignedReverseYPdf);
            });
        });
        it("signs document (no background)", function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete visualInfo.background;
                const noBackgroundVisualSignedPdf = yield pdfSigner.signAsync(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.pdf, visualInfo);
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.paths.noBackgroundVisualSignedPdf, noBackgroundVisualSignedPdf);
                (0, chai_2.expect)(noBackgroundVisualSignedPdf).to.be.deep.equal(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.noBackgroundVisualSignedPdf);
            });
        });
        it("signs document (no texts)", function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete visualInfo.texts;
                const noTextsVisualSignedPdf = yield pdfSigner.signAsync(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.pdf, visualInfo);
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.paths.noTextsVisualSignedPdf, noTextsVisualSignedPdf);
                (0, chai_2.expect)(noTextsVisualSignedPdf).to.be.deep.equal(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.noTextsVisualSignedPdf);
            });
        });
        it("signs document (no optionals)", function () {
            return __awaiter(this, void 0, void 0, function* () {
                delete visualInfo.background;
                delete visualInfo.texts;
                const noOptionalsVisualSignedPdf = yield pdfSigner.signAsync(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.pdf, visualInfo);
                yield _helpers_1.generateAsset.generateBinaryAsync(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.paths.noOptionalsVisualSignedPdf, noOptionalsVisualSignedPdf);
                (0, chai_2.expect)(noOptionalsVisualSignedPdf).to.be.deep.equal(assets_pdf_visual_signer_pdf_1.pdfVisualSignerAssets.noOptionalsVisualSignedPdf);
            });
        });
    });
});
