// sign-pdf-lib/dist/src/signer/pdf-signer.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PdfSigner_digitalSigner, _PdfSigner_visualSigner;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PdfSigner = void 0;
const pdf_digital_signer_1 = require("./pdf-digital-signer");
const pdf_visual_signer_1 = require("./pdf-visual-signer");
class PdfSigner {
    constructor(settings) {
        _PdfSigner_digitalSigner.set(this, void 0);
        _PdfSigner_visualSigner.set(this, void 0);
        __classPrivateFieldSet(this, _PdfSigner_digitalSigner, new pdf_digital_signer_1.PdfDigitalSigner(settings), "f");
        __classPrivateFieldSet(this, _PdfSigner_visualSigner, new pdf_visual_signer_1.PdfVisualSigner(), "f");
    }
    addPlaceholderAsync(pdf, info) {
        return __awaiter(this, void 0, void 0, function* () {
            return __classPrivateFieldGet(this, _PdfSigner_digitalSigner, "f").addPlaceholderAsync(pdf, info);
        });
    }
    addFieldAsync(pdf, info) {
        return __awaiter(this, void 0, void 0, function* () {
            return __classPrivateFieldGet(this, _PdfSigner_digitalSigner, "f").addFieldAsync(pdf, info);
        });
    }
    signAsync(pdf, info) {
        return __awaiter(this, void 0, void 0, function* () {
            return __classPrivateFieldGet(this, _PdfSigner_digitalSigner, "f").signAsync(pdf, info);
        });
    }
    signFieldAsync(pdf, info) {
        return __awaiter(this, void 0, void 0, function* () {
            return __classPrivateFieldGet(this, _PdfSigner_digitalSigner, "f").signFieldAsync(pdf, info);
        });
    }
    signVisualAsync(pdf, info) {
        return __awaiter(this, void 0, void 0, function* () {
            return __classPrivateFieldGet(this, _PdfSigner_visualSigner, "f").signAsync(pdf, info);
        });
    }
    verifySignaturesAsync(pdf) {
        return __awaiter(this, void 0, void 0, function* () {
            return __classPrivateFieldGet(this, _PdfSigner_digitalSigner, "f").verifySignaturesAsync(pdf);
        });
    }
    getFieldsAsync(pdf) {
        return __awaiter(this, void 0, void 0, function* () {
            return __classPrivateFieldGet(this, _PdfSigner_digitalSigner, "f").getFieldsAsync(pdf);
        });
    }
}
exports.PdfSigner = PdfSigner;
_PdfSigner_digitalSigner = new WeakMap(), _PdfSigner_visualSigner = new WeakMap();
