// sign-pdf-lib/dist/src/helpers/rectangle-helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeAbsolutePageRectangle = exports.computeAbsolutePageReverseRectangle = void 0;
const rectangle_1 = require("../models/rectangle");
function computeCoordinate(coordinate, limit) {
    return coordinate >= 0
        ? coordinate
        : (limit + coordinate);
}
function computeAbsolutePageReverseRectangle(relativeRectangle, pageSize) {
    if (!relativeRectangle) {
        return rectangle_1.emptyRectangle;
    }
    return {
        left: computeCoordinate(relativeRectangle.left, pageSize.width),
        top: pageSize.height - computeCoordinate(relativeRectangle.top, pageSize.height),
        right: computeCoordinate(relativeRectangle.right, pageSize.width),
        bottom: pageSize.height - computeCoordinate(relativeRectangle.bottom, pageSize.height)
    };
}
exports.computeAbsolutePageReverseRectangle = computeAbsolutePageReverseRectangle;
function computeAbsolutePageRectangle(relativeRectangle, pageSize) {
    if (!relativeRectangle) {
        return rectangle_1.emptyRectangle;
    }
    return {
        left: computeCoordinate(relativeRectangle.left, pageSize.width),
        top: computeCoordinate(relativeRectangle.top, pageSize.height),
        right: computeCoordinate(relativeRectangle.right, pageSize.width),
        bottom: computeCoordinate(relativeRectangle.bottom, pageSize.height)
    };
}
exports.computeAbsolutePageRectangle = computeAbsolutePageRectangle;
